spec.html

```diff
@@ -5606,6 +5620,25 @@ <h1>Strict Equality Comparison</h1>
   <emu-clause id="sec-operations-on-objects">
     <h1>Operations on Objects</h1>
 
+    <emu-clause id="sec-makebasicobject" aoid="MakeBasicObject">
+      <h1>MakeBasicObject ( _internalSlotsList_ )</h1>
+      <p>The abstract operation MakeBasicObject is the source of all ECMAScript objects that are created algorithmically, including both ordinary objects and exotic objects. It factors out common steps used in creating all objects, and centralizes object creation.</p>
+
+      <emu-alg>
+        1. Assert: _internalSlotsList_ is a List of internal slot names.
+        1. Let _obj_ be a newly created object with an internal slot for each name in _internalSlotsList_.
+        1. Set _obj_'s essential internal methods to the default ordinary object definitions specified in <emu-xref href="#sec-ordinary-object-internal-methods-and-internal-slots"></emu-xref>.
+        1. Assert: If the caller will not be overriding both _obj_'s [[GetPrototypeOf]] and [[SetPrototypeOf]] essential internal methods, then _internalSlotsList_ contains [[Prototype]].
+        1. Assert: If the caller will not be overriding all of _obj_'s [[SetPrototypeOf]], [[IsExtensible]], and [[PreventExtensions]] essential internal methods, then _internalSlotsList_ contains [[Extensible]].
+        1. If _internalSlotsList_ contains [[Extensible]], then set _obj_.[[Extensible]] to *true*.
+        1. Return _obj_.
+      </emu-alg>
+
+      <emu-note>
+        <p>Within this specification, exotic objects are created in abstract operations such as ArrayCreate and BoundFunctionCreate by first calling MakeBasicObject to obtain a basic, foundational object, and then overriding some or all of that object's internal methods. In order to encapsulate exotic object creation, the object's essential internal methods are never modified outside those operations.</p>
```

Comment:
Hm.

[`MakeConstructor`](https://deploy-preview-1460--ecma262-snapshots.netlify.com/#sec-makeconstructor) is passed a function object and sets its `[[Construct]]` internal method, and [`OrdinaryFunctionCreate`](https://deploy-preview-1460--ecma262-snapshots.netlify.com/#sec-ordinaryfunctioncreate) obtains a function object by calling [`OrdinaryObjectCreate`](https://deploy-preview-1460--ecma262-snapshots.netlify.com/#sec-ordinaryobjectcreate) and sets its `[[Call]]` internal method.

I guess those technically are not creating _exotic_ objects, so this is not technically wrong, but I find it misleading (which is to say, it misled me) in light of the above. I believe those are the only two places where the internal slots of an object are manipulated by a method other than the one which invoked `MakeBasicObject`. Perhaps (as a followup) we can refactor things so that those two cases go away, so that this note would apply to all objects (and we could get rid of the "Thus, it is not called by any algorithm that subsequently modifies the internal methods of the object in ways that would make the result non-ordinary." part of the note on `OrdinaryObjectCreate` below, since it would then be redundant).

I'm not asking for changes right now, just noting I found this misleading.

Reply:
> I guess those technically are not creating _exotic_ objects

Right. For `OrdinaryFunctionCreate`, it's clear from the name. For `MakeConstructor`, the first step asserts that _F_ is an ECMAScript (i.e., ordinary) function object.

> but I find it misleading (which is to say, it misled me)

What were you misled to think/expect?

Reply:
> What were you misled to think/expect?

I assumed that the essential internal methods would never be manipulated except by the method which called `MakeBasicObject`. That's not actually true, and is not quite what this line says, but (unless I've missed some other cases) it is very close to being true. And it's what I expected this to say, so on first reading I missed that it says something slightly different.

Reply:
A followup PR sounds good to me.

Reply:
@bakkot: Rather than refactoring to make those two cases go away, an easier alternative (or interim aid) would be to identify them in another Note after OrdinaryObjectCreate (and/or possibly modify the existing one).